src/services/download/http.rs

```diff
@@ -0,0 +1,264 @@
+                        )
+                    })
+                    .and_then(|_| future::ok(dest2));
+                Box::new(fut) as Box<dyn Future<Item = PathBuf, Error = DownloadError>>
+            }
+            None => {
+                let fut = future::err(DownloadError::from(DownloadErrorKind::GenericFailed));
+                Box::new(fut) as Box<dyn Future<Item = PathBuf, Error = DownloadError>>
+            }
+        });
+    Box::new(ret) as Box<dyn Future<Item = PathBuf, Error = DownloadError>>
+}
+// pub fn download_source(
+//     source: Arc<HttpSourceConfig>,
+//     location: SourceLocation,
+//     dest: PathBuf,
+// ) -> Box<dyn Future<Item = PathBuf, Error = DownloadError>> {
+//     // ) -> Box<dyn Future<Item = PathBuf, Error = DownloadError> + Send + 'static> {
+//     // All file I/O in this function is blocking!
+//     let ret = download_from_source(source, &location)
+//         .and_then(|maybe_stream| match maybe_stream {
+//             Some(stream) => future::ok(stream),
+//             None => future::err(DownloadErrorKind::GenericFailed.into()),
+//         })
+//         .map(|stream| {
+//             log::trace!(
+//                 "Downloading file for HTTP source {id} location {loc}",
+//                 id = source.id,
+//                 loc = location
+//             );
+//             let destdir = tryf!(dest
+//                 .parent()
+//                 .ok_or(DownloadError::from(DownloadErrorKind::BadDestination)));
+//             let tmpfile = tryf!(NamedTempFile::new_in(destdir)
+//                 .context(DownloadErrorKind::TempFile)
+//                 .map_err(|e| e.into()));
+//             let tmpfile2 = tryf!(tmpfile
+//                 .reopen()
+//                 .context(DownloadErrorKind::TempFile)
+//                 .map_err(|e| e.into()));
+//             let async_file = tokio::fs::File::from_std(tmpfile2);
+//             let accumulator = stream
+//                 .fold(async_file, |mut file, chunk| {
+//                     file.write_all(chunk.as_ref())
+//                         .context(DownloadErrorKind::Write)
+//                         .map(|_| file)
+//                 })
+//                 .map_err(|err| err.into())
+//                 .and_then(|_| {
+//                     future::result(
+//                         tmpfile
+//                             .persist(dest)
+//                             .context(DownloadErrorKind::Write.into()),
+//                     )
+//                 })
+//                 .map(|_| future::ok(dest));
+//             Box::new(accumulator)
+//         });
+//     // Box::new(ret)
+//     Box::new(ret) as Box<dyn Future<Item = PathBuf, Error = DownloadError>>
+// }
+
+// The actors::objects::http::download_from_source with minimal changes.
+fn download_from_source(
+    source: Arc<HttpSourceConfig>,
+    download_path: &SourceLocation,
+) -> Box<dyn Future<Item = Option<DownloadStream>, Error = DownloadError>> {
```

Comment:
just curious why we have an `Option` here, instead of another `DownloadError` variant.

Reply:
I did wonder that too... this is the bit of code i kept as unchanged as possible.  the glue that calls this transforms it into a DownloadError

Reply:
So this is to not make a 404 an error.  On the service-level interface i've replaced this with a more obvious enum instead.